core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -340,23 +340,48 @@ class GroupCoordinator(val brokerId: Int,
 
   def handleDeleteGroups(groupIds: Set[String]): Map[String, Errors] = {
     if (!isActive.get) {
-      groupIds.map((_ -> Errors.COORDINATOR_NOT_AVAILABLE)).toMap
+      groupIds.map(_ -> Errors.COORDINATOR_NOT_AVAILABLE).toMap
     } else {
       var groupErrors: Map[String, Errors] = Map()
-      var eligibleGroups: Seq[String] = Seq()
+      var eligibleGroups: Seq[GroupMetadata] = Seq()
 
       groupIds.foreach { groupId =>
         if (!validGroupId(groupId))
-          groupErrors += (groupId -> Errors.INVALID_GROUP_ID)
-        else if (!isCoordinatorForGroup(groupId)) {
-          groupErrors += (groupId -> Errors.NOT_COORDINATOR)
-        } else if (isCoordinatorLoadInProgress(groupId))
-          groupErrors += (groupId -> Errors.COORDINATOR_LOAD_IN_PROGRESS)
-        else
-          eligibleGroups ++= Seq(groupId)
+          groupErrors += groupId -> Errors.INVALID_GROUP_ID
+        else if (!isCoordinatorForGroup(groupId))
+          groupErrors += groupId -> Errors.NOT_COORDINATOR
+        else if (isCoordinatorLoadInProgress(groupId))
+          groupErrors += groupId -> Errors.COORDINATOR_LOAD_IN_PROGRESS
+        else {
+          groupManager.getGroup(groupId) match {
+            case None =>
+              groupErrors += groupId -> Errors.GROUP_ID_NOT_FOUND
+            case Some(group) =>
+              group.inLock {
+                group.currentState match {
+                  case Dead =>
+                    if (!isCoordinatorForGroup(groupId))
+                      groupErrors += groupId -> Errors.NOT_COORDINATOR
+                    else
+                      groupErrors += groupId -> Errors.GROUP_ID_NOT_FOUND
+                  case Empty =>
+                    group.transitionTo(Dead)
+                    eligibleGroups :+= group
+                  case _ =>
+                    groupErrors += groupId -> Errors.NON_EMPTY_GROUP
+                }
+              }
+          }
+        }
+      }
+
+      if (eligibleGroups.nonEmpty) {
+        groupManager.cleanupGroupMetadata(None, eligibleGroups, Long.MaxValue)
```

Comment:
This still feels a bit hacky. As an alternative, maybe we can let the offset selector be provided as a function. Something like this:

```scala
def cleanupGroupMetadata(
 groups: Iterable[GroupMetadata], 
 collectOffsetsToRemove: Group => Map[TopicPartition, OffsetAndMetadata])
```

What do you think?

Reply:
I'm not sure which part you consider hacky, and am trying to understand your suggestion.

For the sake of `deleteGroups` functionality, we can use `group.allOffsets` that conforms to the function signature above. But how about the existing functionality, where we want to delete specific topic partitions from a group: `groupManager.cleanupGroupMetadata(Some(topicPartitions), groupManager.currentGroups, time.milliseconds())` and populate the corresponding `OffsetAndMetadata` values? I'm assuming we want to reuse the same `cleanupGroupMetadata` method for both cases.

On the same assumption, we also need to factor in the concept of current time so we can determine the expired offsets for the existing functionality.

On the other hand if you are proposing to create A new `cleanupGroupMetadata` method that calls on the existing method, we should make this call once per group (since topic partitions are group-specific).

Or maybe I'm missing the point :)

Reply:
It's not that big of a deal. I just thought it was a mild abuse to reuse the expiration logic to delete all offsets. Alternatively, what I was suggesting is to let the caller choose the offsets to delete.